translated/tech/Measuring progress and estimating time remaining of long running tasks in Go.md

```diff
@@ -0,0 +1,186 @@
+## 对Go中长时间运行io.Reader和io.Writer的操作测算进度和估算剩余时间
+
+​                                                                   				 		 Mat Ryer
+
+
+
+![img](https://cdn-images-1.medium.com/max/900/1*YfQ0FQIK4l6NMW3wsl9NNw.jpeg)
+
+每当我们在使用类似 io.Copy 和  ioutil.ReadAll 的工具时，比如我们正在从 http.Response 体读入或者上传一个文件，我们会发现这些方法将一直堵塞，直到整个过程完成，哪怕耗时数十分钟甚至是小时——而且我们没有办法来查看进度，以及计算出完成所需剩余时间的估测值。
+
+本文很长，不想深究瞅这里：这篇文章最终会引导到 progress 包，你可以在自己的项目中自由使用——https://github.com/machinebox/progress
+
+考虑到 io.Reader 和 io.Writer 都是接口，我们可以封装它们并且拦截 Read 和 Write 方法，捕获实际已经通过它们的字节数。通过一些简单的数学计算，我们可以计算出已完成部分所占的比例。再多上一点数学计算，我们甚至可以估测整个过程还剩余多少时间，假设传输流是相对一致的话。
+
+## 封装Reader
+
+一个新的 Reader 类型只需要包含另一个 io.Reader , 并且调用它的 Read 方法来获取返回前读到的字节数。为了保证 reader 可以在并发环境中安全使用（在这个例子中至关重要），我们可以使用 atomic.AddInt64 作为安全的计数器。
+
+```go
+// Reader ：计数通过它读取的字节数。
+type Reader struct {
+ r io.Reader
+ n int64
+}
+// NewReader 返回一个可以计数通过它读取到字节数的
+// Reader
+func NewReader(r io.Reader) *Reader {
+ return &Reader{
+ r: r,
+ }
+}
+func (r *Reader) Read(p []byte) (n int, err error) {
+ n, err = r.r.Read(p)
+ atomic.AddInt64(&r.n, int64(n))
+ return
+}
+// N 表示目前为止读取到的字节数
+func (r *Reader) N() int64 {
+ return atomic.LoadInt64(&r.n)
+}
+
+```
+
+试试看你能不在自己写出 Writer 的计数部分，两者很类似。
+
```

Comment:
这里漏了一小段没有翻译：
> Since the N method returns (safely via atomic.LoadInt64) the number of bytes read, we can call this from another goroutine at any time to find out what has happened so far.